services/streammanager/googlecloudfunction/googlecloudfunction.go

```diff
@@ -0,0 +1,226 @@
+	"google.golang.org/api/googleapi"
+	"google.golang.org/api/idtoken"
+
+	"google.golang.org/api/option"
+
+	"github.com/rudderlabs/rudder-go-kit/logger"
+	backendconfig "github.com/rudderlabs/rudder-server/backend-config"
+	"github.com/rudderlabs/rudder-server/services/streammanager/common"
+	"github.com/tidwall/gjson"
+)
+
+type Config struct {
+	Credentials            string `json:"credentials"`
+	FunctionEnvironment    string `json:"functionEnvironment"`
+	RequireAuthentication  bool   `json:"requireAuthentication"`
+	GoogleCloudFunctionUrl string `json:"googleCloudFunctionUrl"`
+}
+
+type Client struct {
+	service *cloudfunctions.Service
+	opts    common.Opts
+}
+
+var pkgLogger logger.Logger
+
+func init() {
+	pkgLogger = logger.NewLogger().Child("streammanager").Child("GoogleCloudFunction")
+}
+
+type GoogleCloudFunctionProducer struct {
+	client *Client
+	config *Config
+}
+
+// NewProducer creates a producer based on destination config
+func NewProducer(destination *backendconfig.DestinationT, o common.Opts) (*GoogleCloudFunctionProducer, error) {
+	var config Config
+	jsonConfig, err := json.Marshal(destination.Config)
+	if err != nil {
+		return nil, fmt.Errorf("[GoogleCloudFunction] Error while marshalling destination config :: %w", err)
+	}
+	err = json.Unmarshal(jsonConfig, &config)
+	if err != nil {
+		return nil, fmt.Errorf("[GoogleCloudFunction] error  :: error in GoogleCloudFunction while unmarshalling destination config:: %w", err)
+	}
+
+	opts := []option.ClientOption{
+		option.WithCredentialsJSON([]byte(config.Credentials)),
+	}
+
+	service, err := generateService(destination, opts...)
+	// If err is not nil then retrun
+	if err != nil {
+		pkgLogger.Errorf("[GoogleCloudFunction] error  :: %w", err)
+		return nil, err
+	}
+	client := &Client{service, o}
+	destConfig := &Config{config.Credentials, config.FunctionEnvironment, config.RequireAuthentication, config.GoogleCloudFunctionUrl}
+
+	return &GoogleCloudFunctionProducer{client, destConfig}, err
+}
+
+func (producer *GoogleCloudFunctionProducer) Produce(jsonData json.RawMessage, _ interface{}) (statusCode int, respStatus, responseMessage string) {
+	destConfig := producer.config
+	parsedJSON := gjson.ParseBytes(jsonData)
+
+	if destConfig.FunctionEnvironment == "gen1" && destConfig.RequireAuthentication {
+
+		return invokeAuthenticatedGen1Functions(producer.client, destConfig.GoogleCloudFunctionUrl, parsedJSON)
+	}
+
+	return invokeGen2AndUnauthenticatedFunctions(destConfig.GoogleCloudFunctionUrl, destConfig.Credentials, destConfig.RequireAuthentication, parsedJSON)
+}
+
+func invokeAuthenticatedGen1Functions(client *Client, functionUrl string, parsedJSON gjson.Result) (statusCode int, respStatus, responseMessage string) {
+	if client == nil {
+		respStatus = "Failure"
+		responseMessage = "[GoogleCloudFunction] error  :: Failed to initialize GoogleCloudFunction client"
+		return 400, respStatus, responseMessage
+	}
+
+	functionName := getFunctionName(functionUrl)
```

Comment:
I have commented earlier to do this in NewProducer to avoid doing this again again during the event processing.

Reply:
This is only being used when we are calling through in the case of gen1 functions. Should we move it to NewProducer?

Reply:
That is fine, as it is done only one time compared to doing for every event